src/v/cluster/id_allocator_frontend.cc

```diff
@@ -0,0 +1,186 @@
+  ss::sharded<cluster::metadata_cache>& metadata_cache,
+  ss::sharded<rpc::connection_cache>& connection_cache,
+  ss::sharded<partition_leaders_table>& leaders,
+  std::unique_ptr<cluster::controller>& controller)
+  : _ssg(ssg)
+  , _partition_manager(partition_manager)
+  , _shard_table(shard_table)
+  , _metadata_cache(metadata_cache)
+  , _connection_cache(connection_cache)
+  , _leaders(leaders)
+  , _controller(controller) {}
+
+ss::future<allocate_id_reply>
+id_allocator_frontend::allocate_id(model::timeout_clock::duration timeout) {
+    auto nt = model::topic_namespace(
+      model::kafka_internal_namespace, model::id_allocator_topic);
+
+    auto has_topic = ss::make_ready_future<bool>(true);
+
+    if (!_metadata_cache.local().contains(nt, model::partition_id(0))) {
+        has_topic = try_create_id_allocator_topic();
+    }
+
+    return has_topic.then([this, timeout](bool does_topic_exist) {
+        if (!does_topic_exist) {
+            return ss::make_ready_future<allocate_id_reply>(
+              allocate_id_reply{0, errc::topic_not_exists});
+        }
+
+        auto leader = _leaders.local().get_leader(model::id_allocator_ntp);
+
+        if (!leader) {
+            vlog(
+              clusterlog.warn,
+              "can't find a leader for {}",
+              model::id_allocator_ntp);
+            return ss::make_ready_future<allocate_id_reply>(
+              allocate_id_reply{0, errc::no_leader_controller});
+        }
+
+        auto _self = _controller->self();
+
+        if (leader == _self) {
+            return do_allocate_id(timeout);
+        }
+
+        vlog(
+          clusterlog.trace,
+          "dispatching allocate id to {} from {}",
+          leader,
+          _self);
+
+        return dispatch_allocate_id_to_leader(leader.value(), timeout);
+    });
+}
+
+ss::future<allocate_id_reply>
+id_allocator_frontend::dispatch_allocate_id_to_leader(
+  model::node_id leader, model::timeout_clock::duration timeout) {
+    return _connection_cache.local()
+      .with_node_client<cluster::id_allocator_client_protocol>(
+        _controller->self(),
+        ss::this_shard_id(),
+        leader,
+        [timeout](id_allocator_client_protocol cp) {
+            return cp.allocate_id(
+              allocate_id_request{timeout},
+              rpc::client_opts(model::timeout_clock::now() + timeout));
+        })
+      .then(&rpc::get_ctx_data<allocate_id_reply>)
+      .then([](result<allocate_id_reply> r) {
+          if (r.has_error()) {
+              vlog(
+                clusterlog.warn,
+                "got error {} on remote allocate id",
+                r.error());
+              return allocate_id_reply{0, errc::timeout};
+          }
+
+          return r.value();
+      });
+}
+
+ss::future<allocate_id_reply>
+id_allocator_frontend::do_allocate_id(model::timeout_clock::duration timeout) {
+    auto shard = _shard_table.local().shard_for(model::id_allocator_ntp);
+
+    if (shard == std::nullopt) {
```

Comment:
no need to change anything: std::nullopt evaluates to false so `if (shard)` is ok.